### DISCUSSION TITLE
Specify custom "folder" name for filename display of virtual documents in VSCode

### DISCUSSION THREAD

**Discussion by wilhelmklopp** [microsoft/vscode-discussions#640] [Extension Development QnA] [answered]
When you open two files in VSCode that have the same filename, VSCode will helpfully display the folder name alongside the filename so that you can tell the two files apart:

[![same file name][1]][1]

In my VSCode extension, I make extensive use of [virtual documents](https://code.visualstudio.com/api/extension-guides/virtual-documents) to display text-based content. However, I can't quite seem to replicate this behaviour of showing the folder name as additional context in the tab.

For example, I have the following two virtual documents:

* The "virtual" path of the first one is `sql/FJ5G91KN/queries.sql`
*  and the "virtual" path of the second one is `sql/6D3B7G5K/queries.sql`

[![virtual documents][2]][2]

However, VSCode does not display those "virtual" folder names (FJ5G91KN and 6D3B7G5K) as additional context in the tab.

How can I get VSCode to show the "virtual" folder name just as it automatically does when you have two regular files opened that have the same filename?

Thanks!

  [1]: https://i.stack.imgur.com/eeECh.png
  [2]: https://i.stack.imgur.com/42Vh7.png

**Comment by r3m0t** [comment]
Are you using the `vscode.open` command or can you share the code that opens these editors? https://code.visualstudio.com/api/references/commands

**Reply by wilhelmklopp** [reply]
I'm using my own `TextDocumentContentProvider` for these virtual documents as described [here](https://code.visualstudio.com/api/extension-guides/virtual-documents), so the code that opens the editors looks something like this:

```ts
    const uri = "sql/6D3B7G5K/queries.sql"
    const doc = await vscode.workspace.openTextDocument(uri);

    await vscode.window.showTextDocument(doc);
    await vscode.languages.setTextDocumentLanguage(doc, "sql");
```

**Reply by r3m0t** [reply]
Try

```
const uri = vscode.Uri.parse(...);
await vscode.commands.executeCommand("vscode.open", uri);
```

**Reply by wilhelmklopp** [reply]
Thanks! I tried that and it results in the same outcome:

<img width="243" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/7718702/9768f3d4-21f3-4a5d-829c-ae3b2e8a6904">

**Reply by r3m0t** [reply]
Well, I tried 😆

**Comment by starball5** [comment]
Related on Stack Overflow: [Specify custom "folder" name for filename display of virtual documents in VSCode](https://stackoverflow.com/q/76439522/11107541)

**Comment by isidorn** [comment]
This is probably simply not supported for virtual documents.

@bpasero can confirm.

**Comment by bpasero** [comment, marked as answer]
I would suggest for virtual documents that should behave and appear like files to rather register a file system provider and use that instead for opening documents. You can contribute the provider to be `readonly` to get the same effect as with virtual documents. 

https://github.com/microsoft/vscode/blob/00039b0fcf281275f6e313bacc830aca31e9bc92/src/vscode-dts/vscode.d.ts#L12756
